### PR TITLE
[SIGNIFICANT] Remove redundant render in react

### PR DESCRIPTION
**Pull request by tmeasday**
Issue:

In the React framework code, `ReactDOM.render` would get unnecessarily called twice.

## What I did

Exit early from the render function if nothing is changing.

## How to test

I added a story "Lifecycle:logging" to the cra-kitchen-sink app that demostrates the problem:

To reproduce: browse to the story, hit reload in the browser (as it happens whenever the preview redux store changes, there are a probably other ways to reproduce).

**Previously**: the component would re-render, which would log extra lines to the console.

**Now**: it simply renders once.

We should also satisfy ourselves that no other feature was relying on this bug (cf https://github.com/storybooks/storybook/pull/2463)

### Is this testable with jest or storyshots?

Not right now; it would be good to maybe use JSDOM to test this kind of thing.

### Does this need a new example in the kitchen sink apps?

Yes, done

### Does this need an update to the documentation?

No

If your answer is yes to any of these, please make sure to include it in your PR.

### CODE REVIEW

File: app/react/src/client/preview/render.js [hunk 1]
```diff
@@ -55,16 +55,18 @@ export function renderMain(data, storyStore) {
...
   //    https://github.com/storybooks/react-storybook/issues/116
-  if (selectedKind !== previousKind || previousStory !== selectedStory) {
-    // We need to unmount the existing set of components in the DOM node.
-    // Otherwise, React may not recrease instances for every story run.
-    // This could leads to issues like below:
-    //    https://github.com/storybooks/react-storybook/issues/81
-    previousKind = selectedKind;
-    previousStory = selectedStory;
-    ReactDOM.unmountComponentAtNode(rootEl);
+  if (selectedKind === previousKind || previousStory === selectedStory) {
```
Comment (igor-dv) [inline app/react/src/client/preview/render.js]: &&?

File: examples/cra-kitchen-sink/src/components/LifecycleLogger.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
+import React, { Component } from 'react';
+
+function log(name) {
+  console.log(`LifecycleLogger: ${name}`);
```
Comment (igor-dv) [inline examples/cra-kitchen-sink/src/components/LifecycleLogger.js]: Feels like a linter warning ?

Comment (tmeasday) [reply examples/cra-kitchen-sink/src/components/LifecycleLogger.js]: Do you think I should use addon-actions for this instead?

File: app/react/src/client/preview/render.js (hunk 1 above)
Comment (tmeasday) [reply app/react/src/client/preview/render.js]: 😊

File: examples/cra-kitchen-sink/src/components/LifecycleLogger.js (hunk 2 above)
Comment (igor-dv) [reply examples/cra-kitchen-sink/src/components/LifecycleLogger.js]: You can just ignore the warning =)

**Comment by igor-dv** [comment]
I wonder if it's easy to add some test that will really check the life cycle.

File: examples/cra-kitchen-sink/src/components/LifecycleLogger.js (hunk 2 above)
Comment (Hypnosphi) [reply examples/cra-kitchen-sink/src/components/LifecycleLogger.js]: You should use `@storybook/client-logger`

Comment (tmeasday) [reply examples/cra-kitchen-sink/src/components/LifecycleLogger.js]: Done.

**Comment by tmeasday** [comment]
@igor-dv 

> I wonder if it's easy to add some test that will really check the life cycle.

Did you have an idea for this? It's tricky because it's testing the top-level of the `app/react` framework package. My thought was maybe using JSDOM and checking the console.logs, but it seems quite complex to implement.

**Comment by tmeasday** [comment]
Any other opinions on this @storybooks/team? Shall we merge it?

**Comment by ndelangen** [comment]
Yeah this makes sense, we want this.

Slightly concerned this will break some plugins, if they depend on this behavior.

But let's merge and fix the addons if it happens. Related: https://github.com/storybooks/storybook/pull/2463

**Review by ndelangen — APPROVED** [review]

**Pull request merged**